tests/0151-purge-brokers.c

```diff
@@ -0,0 +1,410 @@
+        if (!rd_atomic32_get(&do_test_remove_then_add_received_terminate) &&
+            strstr(buf, "/1: Handle terminates in state") > 0) {
+                rd_atomic32_set(&do_test_remove_then_add_received_terminate, 1);
+                while (!rd_atomic32_get(&verification_complete))
+                        rd_usleep(100 * 1000, 0);
+        }
+}
+
+/**
+ * @brief Await for the TERMINATE op to be received after the action
+ *        that removes the broker then proceed to
+ *        add the broker again.
+ */
+static rd_bool_t do_test_remove_then_add_await_after_action_cb(int action) {
+        if (action == 2) {
+                /* Wait until TERMINATE is received */
+                return !rd_atomic32_get(
+                    &do_test_remove_then_add_received_terminate);
+        }
+        return rd_false;
+}
+
+/**
+ * @brief Disable sparse connections to increase likely of problems
+ *        when the decommisioned broker is re-connecting.
+ *        Add a pause after receiving the TERMINATE op to allow to
+ *        proceed with adding it again before it's decommissioned.
+ */
+static test_conf_log_interceptor_t *log_interceptor;
+static void
+do_test_remove_then_add_edit_configuration_cb(rd_kafka_conf_t *conf) {
+        const char *debug_contexts[2] = {"broker", NULL};
+
+        /* This timeout verifies that the correct brokers are returned
+         * without duplicates as soon as possible. */
+        test_timeout_set(6);
+        /* Hidden property that forces connections to all brokers,
+         * increasing likelyhood of wrong behaviour if the decommissioned broker
+         * starts re-connecting. */
+        test_conf_set(conf, "enable.sparse.connections", "false");
+        log_interceptor = test_conf_set_log_interceptor(
+            conf, do_test_remove_then_add_log_cb, debug_contexts);
+}
+
+/**
+ * @brief Test setting down one broker and then adding it again
+ *        while it's still being decommissioned.
+ *
+ *        This should not leave dangling references that prevent broker
+ *        destruction.
+ */
+static void do_test_remove_then_add(void) {
+        SUB_TEST_QUICK();
+        rd_atomic32_init(&do_test_remove_then_add_received_terminate, 0);
+        rd_atomic32_init(&verification_complete, 0);
+
+        int32_t expected_brokers_cnt[] = {3, 3, 2, 3};
+
+        int32_t expected_broker_ids[][5] = {
+            {1, 2, 3}, {1, 2, 3}, {2, 3}, {1, 2, 3}};
+
+        int32_t actions[][2] = {
+            {TEST_ACTION_SET_GROUP_COORDINATOR, 3},
+            {TEST_ACTION_REMOVE_BROKER, 1},
+            {TEST_ACTION_ADD_BROKER, 1},
+        };
+
+        do_test_add_remove_brokers(
+            3, actions, expected_broker_ids, RD_ARRAY_SIZE(expected_broker_ids),
+            expected_brokers_cnt, do_test_remove_then_add_edit_configuration_cb,
+            do_test_remove_then_add_await_after_action_cb);
+
+        rd_free(log_interceptor);
+        log_interceptor = NULL;
```

Comment:
Instead of global variable, we should get this from the `rk` opaque. `do_test_add_remove_brokers` can return this.

Reply:
ok, just I return it as `void *`  to avoid making it coupled with this test.